### PR TITLE
[SPARK-37933][SQL] Limit push down for parquet vectorized reader

### PR DESCRIPTION
**Pull request by jackylee-ch**
### Why are the changes needed?
Based on [34291](https://github.com/apache/spark/pull/34291), we can support limit push down to parquet datasource v2 reader, which can stop scanning parquet early, and reduce network and disk IO.
Currently, only vectorized reader is supported in this pr. Row based reader with limit pushdown needs to be supported in parquet-hadoop first, thus it will be supported in the next pr.

Limit parse status for parquet
before
```
== Physical Plan ==
CollectLimit 10
+- *(1) ColumnarToRow
   +- BatchScan[a#0, b#1] ParquetScan DataFilters: [], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/datasources.db/test_push_down/par..., PartitionFilters: [], PushedAggregation: [], PushedFilters: [], PushedGroupBy: [], ReadSchema: struct<a:int,b:int>, PushedFilters: [], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: [] 
```
After
```
== Physical Plan ==
CollectLimit 10
+- *(1) ColumnarToRow
   +- BatchScan[a#0, b#1] ParquetScan DataFilters: [], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/datasources.db/test_push_down/par..., PartitionFilters: [], PushedAggregation: [], PushedFilters: [], PushedGroupBy: [], ReadSchema: struct<a:int,b:int>, PushedFilters: [], PushedAggregation: [], PushedGroupBy: [], PushedLimit: Some(10) RuntimeFilters: [] 
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
origin tests and new tests

### CODE REVIEW

**Comment by stczwd** [comment]
cc @huaxingao @cloud-fan

**Comment by stczwd** [comment]
also cc @sunchao

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 1]
```diff
@@ -139,14 +142,34 @@ public VectorizedParquetRecordReader(
...
+    this.datetimeRebaseTz = datetimeRebaseTz;
+    this.int96RebaseMode = int96RebaseMode;
+    this.int96RebaseTz = int96RebaseTz;
+    MEMORY_MODE = useOffHeap ? MemoryMode.OFF_HEAP : MemoryMode.ON_HEAP;
+    this.capacity = capacity;
+    this.limit = limit;
+  }
+
+  public VectorizedParquetRecordReader(
+          ZoneId convertTz,
```
Comment (huaxingao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:158]: 4-space indentation

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 2]
```diff
@@ -302,10 +325,13 @@ public boolean nextBatch() throws IOException {
...
     columnarBatch.setNumRows(0);
-    if (rowsReturned >= totalRowCount) return false;
+    if (rowsReturned >= totalRowCount || rowsReturned >= limit) {
+      return false;
+    }
     checkEndOfRowGroup();
 
-    int num = (int) Math.min((long) capacity, totalCountLoadedSoFar - rowsReturned);
+    int num = (int) Math.min((long) capacity,
+            Math.min((long) limit - rowsReturned, totalCountLoadedSoFar - rowsReturned));
```
Comment (huaxingao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:334]: indentation is off

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetLimitPushDownSuite.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      SQLConf.PARQUET_LIMIT_PUSHDOWN_ENABLED.key -> "true",
+      SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> "true",
+      SQLConf.USE_V1_SOURCE_LIST.key -> "") {
+      withTempPath { path =>
+        (1 to 1024).map(i => (101, i)).toDF("a", "b").coalesce(1).write.parquet(path.getPath)
+        val pushedLimit = Random.nextInt(100)
+        val df = spark.read.parquet(path.getPath).limit(pushedLimit)
+        val sparkPlan = df.queryExecution.sparkPlan
+        sparkPlan foreachUp  {
+          case r@ BatchScanExec(_, f: ParquetScan, _) =>
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetLimitPushDownSuite.scala:45]: space between `r` and `@`

**Comment by huaxingao** [comment]
@stczwd Thanks for working on this! The changes look reasonable to me. I left a couple of nit comments for coding style.

**Review by sunchao — COMMENTED** [review]
Thanks @stczwd .

I'm curious whether this will be very useful. To my knowledge, Spark already pushes local limit to each task, see `LimitExec.doExecute`:
```scala
  protected override def doExecute(): RDD[InternalRow] = {
    val childRDD = child.execute()
    if (childRDD.getNumPartitions == 0) {
      new ParallelCollectionRDD(sparkContext, Seq.empty[InternalRow], 1, Map.empty)
    } else {
      val singlePartitionRDD = if (childRDD.getNumPartitions == 1) {
        childRDD
      } else {
        val locallyLimited = childRDD.mapPartitionsInternal(_.take(limit))
        new ShuffledRowRDD(
          ShuffleExchangeExec.prepareShuffleDependency(
            locallyLimited,
            child.output,
            SinglePartition,
            serializer,
            writeMetrics),
          readMetrics)
      }
      singlePartitionRDD.mapPartitionsInternal(_.take(limit))
    }
  }
```

Since the vectorized Parquet reader behaves like an iterator of `ColumnarBatch`es, it will stop reading more batches when the local limit is reached. 

Also have you done any benchmark with this feature?

**Comment by stczwd** [comment]
Thanks for your reply, @sunchao.
Limit pushdown can terminate the capacity batch reading in advance, it is used especially when reading across rowgroups, which can reduce the reading of one rowgroup.
I have test `limitBenchMark` in `ParquetNestedSchemaPruningBenchmark`. Unfortunately it didn't show clear advantage.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java (hunk 1 above)
Comment (stczwd) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: fixed

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 4]
```diff
@@ -139,14 +142,34 @@ public VectorizedParquetRecordReader(
...
+          boolean useOffHeap,
+          int capacity) {
     this.convertTz = convertTz;
     this.datetimeRebaseMode = datetimeRebaseMode;
     this.datetimeRebaseTz = datetimeRebaseTz;
     this.int96RebaseMode = int96RebaseMode;
     this.int96RebaseTz = int96RebaseTz;
     MEMORY_MODE = useOffHeap ? MemoryMode.OFF_HEAP : MemoryMode.ON_HEAP;
     this.capacity = capacity;
+    this.limit = Integer.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (stczwd) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: fixed

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java (hunk 2 above)
Comment (stczwd) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: fixed

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetLimitPushDownSuite.scala (hunk 3 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetLimitPushDownSuite.scala]: fixed

**Comment by c21** [comment]
I have same concern with @sunchao. IMO the improvement might not be significant as we already have limit to not read extra batch during execution. On the other hand, the DSv2 ORC vectorized reader does not have control of skipping row group, as Parquet reader, so the limit push down cannot be implemented for ORC reader in the future. And here we have to introduce code complexity in the common interface `FileScanBuilder`. I feel this is an overkill unless we can showcase some obvious improvements on current benchmark.

**Comment by stczwd** [comment]
Thanks for your reply @c21. 
Sorry, my last test didn't adjust the limit amount, so the limit didn't really work. 
This time I also test `limitBenchMark` in `ParquetNestedSchemaPruningBenchmarkset`, but set the batch capacity to 10240 and the limit to 12500. Now we found that the performance improved by 1.3x.
```
OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Linux 5.11.0-1025-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Limiting:                                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Top-level column with out limit                      83            101          21         12.1          82.7       1.0X
Nested column with out limit                         77             95          11         12.9          77.4       1.1X
Nested column in array with out limit               105            121          19          9.5         105.1       0.8X
Top-level column with limit                          61             69           5         16.3          61.3       1.3X
Nested column with limit                             66             73           7         15.2          65.8       1.3X
Nested column in array with limit                   101            113          20          9.9         101.2       0.8X
```

**Comment by stczwd** [comment]
In essence, the improvement of this pr is to reduce the amount of data obtained for the last time, especially in the following scenarios.
1. the capacity is set relatively large, but with limit we only need to read a small batch.
3. the last batch need to query 2 rowgroups, but with limit we only need to read only one rowgroup.

**Comment by LuciferYang** [comment]
Can you resolve the conflict first @stczwd

**Pull request closed without merging**